lab2 numpy pandas và matplotlib nang cao

1.numpy

In [2]:
import numpy as np 
x= np.array ([1,2,3,4,5,6,7,8,])
#lấy số 2:
x[1]


np.int64(2)

In [6]:
#lấy từ 1 tới 3
x[:3]


array([1, 2, 3])

In [7]:
#lấy từ 3 tới 8
x[2:8]
x[2:]


array([3, 4, 5, 6, 7, 8])

In [11]:
#lay index 2,4
x[[2,4]]



array([3, 5])

In [12]:
data=[[1,2],[3,4],[5,6]]
mang=np.array(data)
mang

array([[1, 2],
       [3, 4],
       [5, 6]])

In [13]:
mang[0,1]

np.int64(2)

In [15]:
mang[-1,1]
mang[2,1]

np.int64(6)

In [31]:
#lay 3,4,5,6
mang[1:, :]
mang[1:]
mang[[1,2]]
mang[1:3]
mang[-2:]


array([[3, 4],
       [5, 6]])

In [32]:
#lay 1,3,5
mang[:,0]
mang[[0,2],0]

array([1, 5])

In [33]:
#lay1 va 5 theo fortmat start end step
mang[::2,0]
mang[0::2,0]

array([1, 5])

In [35]:
x=[
    [1,2,3,4],
    [5,6,7,8],
    [9,10,11,12],
    [13,14,15,16]
]
y=np.array(x)
y


array([[ 1,  2,  3,  4],
       [ 5,  6,  7,  8],
       [ 9, 10, 11, 12],
       [13, 14, 15, 16]])

In [42]:
y[:,1]

array([ 2,  6, 10, 14])

In [44]:
#lay 4,7,10
y[[0, 1, 2], [3, 2, 1]]

array([ 4,  7, 10])

In [59]:
#lay 16, 15, 14, 13
y[3, :: -1]
y[-1 ,-1::-1]


array([16, 15, 14, 13])

bài tập tết về nhà :\
cho ma trận =x=[\
    [99,99,99],\
    [99,99,99],\
    [99,99,99]\
]\
giả sử 0=o và 1 là x\
nhận đầu vào từ phía x và o luân phiên \
cho một cặp chỉ số, nếu phía x nhập ((0,0)) thì ma trận trở thành:\
[\
    [X,99,99],\
    [99,99,99],\
    [99,99,99]\
]\
nếu phía o nhập ((0,0)) thì yêu cầu nhập lại và nếu không thì điền vào ma trận \
nếu ai đó có 3 ô liên tiếp thì dừng trò chơi 



In [4]:
def in_ban_co(ban_co):
    for hang in ban_co:
        print(" | ".join(hang))
        print("-" * 9)

def kiem_tra_nguoi_thang(ban_co, nguoi_choi):
    # Kiểm tra hàng, cột và đường chéo
    for i in range(3):
        if all([o == nguoi_choi for o in ban_co[i]]):
            return True
        if all([ban_co[j][i] == nguoi_choi for j in range(3)]):
            return True
    if all([ban_co[i][i] == nguoi_choi for i in range(3)]) or all([ban_co[i][2-i] == nguoi_choi for i in range(3)]):
        return True
    return False

def caro():
    ban_co = [['99' for _ in range(3)] for _ in range(3)]
    nguoi_choi = ['X', 'O']
    luot_choi = 0

    while True:
        in_ban_co(ban_co)
        nguoi_hien_tai = nguoi_choi[luot_choi % 2]
        print(f"Lượt của người chơi {nguoi_hien_tai}")
        try:
            hang, cot = map(int, input("Nhập hàng và cột (0, 1, hoặc 2): ").split())
        except ValueError:
            print("Đầu vào không hợp lệ, vui lòng nhập hai số cách nhau bởi dấu cách.")
            continue

        if hang not in range(3) or cot not in range(3):
            print("Chỉ số không hợp lệ, vui lòng nhập lại.")
            continue

        if ban_co[hang][cot] != '99':
            print("Nước đi không hợp lệ, thử lại.")
            continue

        ban_co[hang][cot] = nguoi_hien_tai

        if kiem_tra_nguoi_thang(ban_co, nguoi_hien_tai):
            in_ban_co(ban_co)
            print(f"Người chơi {nguoi_hien_tai} thắng!")
            break

        if all(o != '99' for hang in ban_co for o in hang):
            in_ban_co(ban_co)
            print("Hòa!")
            break

        luot_choi += 1

caro()


99 | 99 | 99
---------
99 | 99 | 99
---------
99 | 99 | 99
---------
Lượt của người chơi X
99 | 99 | 99
---------
99 | X | 99
---------
99 | 99 | 99
---------
Lượt của người chơi O
99 | 99 | 99
---------
99 | X | 99
---------
99 | O | 99
---------
Lượt của người chơi X
99 | 99 | 99
---------
99 | X | X
---------
99 | O | 99
---------
Lượt của người chơi O
Đầu vào không hợp lệ, vui lòng nhập hai số cách nhau bởi dấu cách.
99 | 99 | 99
---------
99 | X | X
---------
99 | O | 99
---------
Lượt của người chơi O
Nước đi không hợp lệ, thử lại.
99 | 99 | 99
---------
99 | X | X
---------
99 | O | 99
---------
Lượt của người chơi O
99 | 99 | 99
---------
99 | X | X
---------
99 | O | O
---------
Lượt của người chơi X
99 | 99 | 99
---------
X | X | X
---------
99 | O | O
---------
Người chơi X thắng!


cho ma trận :\
1: lấy 4,5,6\
2: lấy 3,5\
3: lấy 3,4\
lấy 9,6,3


In [5]:
x=[
    [1,2,3],
    [4,5,6],
    [7,8,9]
]
y = np.array(x)
y

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

In [4]:
#1: lấy 4,5,6
y[1]

array([4, 5, 6])

In [6]:
#2: lấy 2,5
y[[0, 1], [1,1]]

array([2, 5])

In [7]:
#3: lấy 3,4
y[[0, 1], [2,0]]

array([3, 4])

In [8]:
#lấy 9,6,3
y[:3,-1][::-1]
y[::-1,2]

array([9, 6, 3])

In [ ]:
z=[1,2,3,4,5,6,7,8,9,10]

q=np.array(z)
q

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

In [11]:
#lay ra cac phan tu mang chan
q= [num for num in z if num%2==0]
q

[2, 4, 6, 8, 10]

In [12]:
np.ones((3,3))

array([[1., 1., 1.],
       [1., 1., 1.],
       [1., 1., 1.]])

In [13]:
x=np.arange(3)
x.shape

(3,)

In [14]:
np.ones((3,3))+np.arange(3)

array([[1., 2., 3.],
       [1., 2., 3.],
       [1., 2., 3.]])

In [15]:
x.reshape ((3,1))

array([[0],
       [1],
       [2]])

In [16]:
x+y

array([[ 1,  3,  5],
       [ 4,  6,  8],
       [ 7,  9, 11]])

tạo một mảng numpy có kích thước 150x5\
hãy tưởng tượng mảng này chứa 150 mẫu về chiều cao, cân nặng , tuổi, lương, gpa của sinh viên Văn Lang\
chia mảng 4 cột đầu tiên thành 1 biến có tên là x và cột cuối cùng là y\
chia x thành x_train và x_test chứa 70% dữ liệu và chia y thành Y_train và Y_test, trong đó y_train chứa 70% dữ liệu .\
tạo 10 tập dữ liệu không chồng chéo của x_train

In [60]:
data=np.random.uniform(0,10,(150,5))
data.shape

(150, 5)

In [61]:
x= data[:,:-1]
y=data[:,-1]
x.shape, y.shape

((150, 4), (150,))

In [63]:
dataset_size=x.shape[0]
dataset_size

150

In [64]:
train_size=int(0.7*dataset_size)
train_size

105

In [ ]:
xtrain= x[:train_size]
xtest= x[train_size:]

xtrain.shape, xtest.shape

((105, 4), (45, 4))

In [68]:
split= xtrain.shape[0]//10
split

10

In [74]:
for counter in range (0,xtrain.shape [0], split):
    print(xtrain[counter: counter + split].shape)

(10, 4)
(10, 4)
(10, 4)
(10, 4)
(10, 4)
(10, 4)
(10, 4)
(10, 4)
(10, 4)
(10, 4)
(5, 4)


**input**
- mảng numpy chứa 150 mẫu về chiều cao, cân nặng , tuổi, lương, gpa của sinh viên Văn Lang.
- 4 cột đầu tiên thành 1 biến có tên là x và cột cuối cùng là y x(150,4) y(150,1).
- Chia dữ liệu thành tập train (70%) và test (30%).
- Chia x_train thành 10 tập dữ liệu con không chồng chéo.





**output**
- mảng numpy được tạo với kích thước 150*5.
- Kích thước của x_train: (105, 4) (chứa 105 mẫu, 4 đặc trưng).
- Kích thước của x_test: (45, 4) (chứa 45 mẫu, 4 đặc trưng).
- Tổng số tập dữ liệu con trong x_train: 10 tập.
- Mỗi tập dữ liệu con trong x_train có kích thước: (10, 4) và dư 1 tập có giá trị (5, 4 ).

**nhận xét**
- Dữ liệu được chia theo tỷ lệ 70%-30% đảm bảo tập train lớn hơn tập test, giúp mô hình học tốt hơn.
- xtrain được chia thành 10 tập con có kích thước (10, 4) phù hợp cho một số kỹ thuật trong học máy giúp kiểm tra mô hình tốt hơn.
- vì 10 tập dữ liệu không chồng chéo nên các tập con của xtrain không bị chồng lấn, đảm bảo tính độc lập khi sử dụng trong huấn luyện mô hình.

